### PR TITLE
Fixed Security::token

### PR DESCRIPTION
**Pull request by sergeyklay**
Fixed `Security::token`
- Improved hash algo
- Added random length and padding params
- Removed all unneeded chars

Now `Security::token` result may contain A-Z, a-z, 0-9, `-` and `_`. `=` is also used if `$padding` is true.

See [RFC 3548](http://tools.ietf.org/html/rfc3548) for the definition of URL-safe base64.

### CODE REVIEW

File: classes/Kohana/Security.php [hunk 1]
```diff
@@ -33,12 +38,15 @@ class Kohana_Security {
 	 *     ));
 	 *
 	 * This provides a basic, but effective, method of preventing CSRF attacks.
+	 * If $n is not specified, Secure::RANDOM_LENGTH is assumed. It may be larger in future.
 	 *
-	 * @param   boolean $new    force a new token to be generated?
-	 * @return  string
+	 * @param   boolean $new     force a new token to be generated?
+	 * @param   integer $n       specifies the length of the random string
+	 * @param   bool    $padding
```
Comment (enov) [inline classes/Kohana/Security.php:45]: I am not sure about the importance of this parameter.

File: classes/Kohana/Security.php [hunk 2]
```diff
@@ -33,12 +38,15 @@ class Kohana_Security {
...
+	 * @return  string  The result may contain A-Z, a-z, 0-9, "-" and "_". "=" is also used if $padding is true.
 	 * @uses    Session::instance
 	 */
-	public static function token($new = FALSE)
+	public static function token($new = FALSE, $n = NULL, $padding = FALSE)
```
Comment (enov) [inline classes/Kohana/Security.php:49]: Could you please set the default `$n = 16` here? Why did you choose 16 (it was openssl_random_pseudo_bytes(32))?

Comment (enov) [inline classes/Kohana/Security.php:49]: No need for Security::RANDOM_LENGTH constant

File: classes/Kohana/Security.php [hunk 3]
```diff
@@ -52,14 +60,17 @@ public static function token($new = FALSE)
 			{
 				// Generate a random pseudo bytes token if openssl_random_pseudo_bytes is available
 				// This is more secure than uniqid, because uniqid relies on microtime, which is predictable
-				$token = base64_encode(openssl_random_pseudo_bytes(32));
+				$token = pack('a*', openssl_random_pseudo_bytes($n ?: Security::RANDOM_LENGTH));
+				$token = str_replace(array("\n", "\r", "\n\r"), '', $token);
```
Comment (enov) [inline classes/Kohana/Security.php:64]: Why are you removing those newline characters? Wouldn't `base64_encode` take care of them, make them URL safe?

File: classes/Kohana/Security.php [hunk 4]
```diff
@@ -52,14 +60,17 @@ public static function token($new = FALSE)
...
 			}
 			else
 			{
 				// Otherwise, fall back to a hashed uniqid
-				$token = sha1(uniqid(NULL, TRUE));
+				$token = substr(hash('sha256', uniqid(null, true)), 0, $n ?: Security::RANDOM_LENGTH);
 			}
 
+			$token = $padding ? strtr(base64_encode($token), '+/', '-_') : rtrim(strtr(base64_encode($token), '+/', '-_'), '=');
```
Comment (enov) [inline classes/Kohana/Security.php:72]: I don't think giving the choice whether to keep the '=' padding or not is important here.

From the RFC you linked:

>   In some circumstances, the use of padding ("=") in base encoded data
>    is not required nor used.  In the general case, when assumptions on
>    size of transported data cannot be made, padding is required to yield
>    correct decoded data.

I suggest you either keep it or trim it.

File: classes/Kohana/Security.php [hunk 5]
```diff
@@ -83,25 +94,25 @@ public static function check($token)
 	{
 		return Security::slow_equals(Security::token(), $token);
 	}
-
```
Comment (enov) [inline classes/Kohana/Security.php:86]: Kindly do not submit formatting changes with this PR. Make them separate PR if those changes are important.

**Comment by enov** [comment]
@sergeyklay Thanks for this. I reviewed and added notes. Thank you for your time.

If you have more time to work on this, I suggest you enhance it further to make token generation decoupled from session and PR anew to 3.4.

Thanks again.

**Comment by WinterSilence** [comment]
**Also, token must be longer than 2 characters**

**Comment by enov** [comment]
bump @sergeyklay . Are you still willing to work on this?

**Comment by sergeyklay** [comment]
Okay. Give me some time

**Comment by sergeyklay** [comment]
Unfortunately I do not have time to do this
